active/0000-serve-file-api.md

```diff
@@ -0,0 +1,149 @@
+
+In order to mitigate the need to diverge into another middleware which behaves almost same as `broccoli-middleware`, this RFC proposes to split the work of setting the headers and serving the files via `broccoli-middleware` and expose a public API that will allow an addon to define how it wants to serve the assets. It will be a low level public API that will be invoked by certain addons.
+
+# Detailed design
+
+Currently the [`serve-files`](https://github.com/ember-cli/ember-cli/blob/375f3a32f4564465d2eccc3815cb61b570ce29f0/lib/tasks/server/middleware/serve-files/index.js#L18) addon defines how it will serve the incoming asset requests. It invokes the `broccoli-middleware` which is responsible for three sets of things:
+
+1. Setting the response headers
+
+2. serving the files and ending the response
+
+3. Shows an error template if it is a build error
+
+## Public API
+This RFC proposes expose two new in-repo addons in `ember-cli` which will now split the above work and remove the `serve-files` addon:
+
+1. `ember-cli:broccoli:watcher`: This addon will contain the middleware which will be responsible for making sure the build is done and will set the response headers that `broccoli-middleware` is doing today. After setting the response headers, it will call the next middleware in the chain. In addition, if the build results in an error, it will show the error template and not terminate the response.
+
+2. `ember-cli:broccoli:serve-files`: This addon will always run *after* `ember-cli:broccoli:watcher` addon. It will contain a middleware that will be responsible for serving the files from the filesystem and ending the response.
+
+For any infrastructure that needs to serve the assets in its own way will be create an addon that will be injected between the above two addons. It will use the `serverMiddleware` public hook to provide its own middleware. Specifically the custom addon should run *before* `ember-cli:broccoli:serve-files` so that it can either override any response headers or can serve the files using its own logic and end the response. This will ensure that when the build is successful `ember-cli:broccoli:watcher` can call the correct next middleware in the chain.
+
+## Implementation Details
+In order for the above API to be exposed, we need to drop the `serve-files` addon in `ember-cli`, refactor `broccoli-middleware` and create the two new addons.
+
+### Refactor `broccoli-middleware` to be a utility
+`broccoli-middleware` is currently responsible for setting the response headers and serving the files. It is a middleware that does these two tasks. It doesn't expose a proper API to do the two tasks differently. We would like to refactor `broccoli-middleware` to be a utility such that it exposes three different APIs:
```

Comment:
A further refinement that utilizes the existing middleware ecosystem, and prevents any new concepts from being added would be as follows:

from the point of view of ember-cli, I'm not sure this matters. I believe ember-cli should ask broccoli-middleware for the two middleware. The reason for this, is to ensure we don't add additional concepts, when the existing concepts suffice. In addition, the express middleware ecosystem is what we use, and we should continue to maintain and improve compatibility with the rest of the ecosystem. 


it needs:

* watcher middleware
* serve files middelware

## watcher middleware

```js
require('broccoli-middleware').forWatcher(watcher);
```
Implemented something like:

```js
function forWatcher(watcher) {
  return function middleware(request, response, next) {
     // content the same as from: https://github.com/ember-cli/rfcs/pull/80#pullrequestreview-9709454
  }
}
```

## Serve Files
```js
require('broccoli-middleware').serveFiles;
```

Which is basically as described in: https://github.com/ember-cli/rfcs/pull/80#pullrequestreview-9709454

---

These two middleware can then be inserted in the existing express middleware chain, and if plugin authors wish to augment, they have that extension point.

---

Now, to implement the existing middleware (for backwards compat), we can merely have 1 middleware that in-effect invokes the above 2 in sequence